File: DevEnvironment/ClientServer/src/main/java/clientserver/core/RetryHandler.java
package clientserver.core;

import base.core.Conditions;
import base.interfaces.ILogic;
import clientserver.interfaces.IRetryLogic;
import clientserver.interfaces.IRetryHandler;
import clientserver.interfaces.IRetryPolicy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The RetryHandler class implements a retry handler.
 */
public final class RetryHandler implements IRetryHandler {
    private final IRetryPolicy policy;
    private final ILogic idleLogic;

    private final Logger log = LoggerFactory.getLogger(this.getClass());

    /**
     * The RetryHandler constructor.
     */
    public RetryHandler() {
        this(
            RetryPolicy.defaultRetryPolicy(),
            null);
    }

    /**
     * The RetryHandler constructor.
     */
    public RetryHandler(IRetryPolicy policy, ILogic idleLogic) {
        Conditions.validateNotNull(
            policy,
            "The retry policy.");

        this.policy = policy;

        if (idleLogic != null) {
            this.idleLogic = idleLogic;
        } else {
            this.idleLogic = new IdleHandler(policy.getIntervalDuration());
        }
    }

    /**
     * Retries to run a logic.
     */
    @Override
    public void retry(IRetryLogic logic) {
        Conditions.validateNotNull(
            logic,
            "The retry logic for running.");

        int attemptIndex = 0;

        String logicInformation = this.getLogicInformation(logic);

        String informationalMessage = "Retry Handler started running for " + logicInformation;
        this.log.info(informationalMessage);

        while (attemptIndex <= this.policy.getNumberOfAttempts()) {
            ++attemptIndex;

            String attemptInformation = this.getRetryLogicAttemptInformation(logic, attemptIndex);

            informationalMessage = "Running " + attemptInformation;
            this.log.info(informationalMessage);

            try {
                logic.run();

                informationalMessage = "Successful " + attemptInformation;
                this.log.info(informationalMessage);

                return;
            } catch (RetryException e) {
                String warningMessage = "Failed " + attemptInformation + ", with error: " + e.getMessage();
                this.log.warn(warningMessage);

                if (attemptIndex < this.policy.getNumberOfAttempts() && this.isAbsorbedException(logic, e)) {
                    this.idleLogic.run();

                    continue;
                }

                String errorMessage = "Retry Handler failed running: " + logicInformation;
                this.log.error(errorMessage);

                throw e;
            }
        }
    }

    /**
     * Checks whether an exception is absorbed.
     */
    private boolean isAbsorbedException(IRetryLogic logic, Exception exception) {
        for (Class<?> absorbedExceptionType : logic.getAbsorbedExceptions()) {
            if (absorbedExceptionType.isAssignableFrom(exception.getClass())) {
                return true;
            }
        }

        return false;
    }

    /**
     * Gets the information of a retry logic.
     */
    private String getLogicInformation(IRetryLogic logic) {
        return "Logic: " + logic + ", with Retry Policy: " + this.policy;
    }

    /**
     * Gets the information of a retry attempt.
     */
    private String getRetryLogicAttemptInformation(IRetryLogic logic, int attemptIndex) {
        return
            "Attempt: " + attemptIndex + "/" + this.policy.getNumberOfAttempts() +
            " of Logic: " + logic;
    }
}

File: SystemCore/ClientServer/core/RetryHandler.cpp
#include "RetryHandler.h"
#include "IRetryLogic.h"
#include "RetryPolicy.h"
#include "RetryException.h"
#include "SmartPointers.h"
#include "ErrorMessages.h"
#include <sstream>

using namespace client_server;

/**
 * Creates a default retry handler.
 */
IRetryHandlerSharedPtr RetryHandler::MakeDefault()
{
    IRetryPolicySharedPtr policy = RetryPolicy::MakeDefault();
    return Make(policy, nullptr);
}

/**
 * Creates a RetryHandler.
 */
IRetryHandlerSharedPtr RetryHandler::Make(
    IRetryPolicySharedPtr policy,
    ILogicSharedPtr idleLogic)
{
    return std::make_shared<RetryHandler>(policy, idleLogic);
}

/**
 * The ConsolePipe constructor.
 */
RetryHandler::RetryHandler(
    IRetryPolicySharedPtr policy,
    ILogicSharedPtr idleLogic)
{
    base::SmartPointers::Validate(policy);
    base::SmartPointers::Validate(idleLogic);

    m_policy = policy;
    m_idleLogic = idleLogic;
}

/**
 * The RetryHandler destructor.
 */
RetryHandler::~RetryHandler()
{
}

/**
 * Retries to run a logic.
 */
void RetryHandler::Retry(IRetryLogic& logic)
{
    unsigned int attemptIndex = 0;

    std::wstring logicInformation = GetLogicInformation(logic);

    if (m_outstream != nullptr) {
        std::wstringstream informationalMessageStream;
        informationalMessageStream << L"Retry Handler started running for " << logicInformation;
        *m_outstream << informationalMessageStream.str();
    }

    unsigned int numberOfAttempts = m_policy->GetNumberOfAttempts();

    while (attemptIndex <= numberOfAttempts) {
        ++attemptIndex;

        if (m_outstream != nullptr) {
            std::wstring attemptInformation = GetRetryLogicAttemptInformation(logic, attemptIndex);
            std::wstringstream informationalMessageStream;
            informationalMessageStream << L"Running " << attemptInformation;

            *m_outstream << informationalMessageStream.str();
        }

        try {
            logic.Run();

            if (m_outstream != nullptr) {
                std::wstring attemptInformation = GetRetryLogicAttemptInformation(logic, attemptIndex);
                std::wstringstream informationalMessageStream;
                informationalMessageStream << L"Successful " << attemptInformation;

                *m_outstream << informationalMessageStream.str();
            }

            return;
        } catch (RetryException& e) {
            long errorCode = e.GetStatusCode();
            std::wstring attemptInformation = GetRetryLogicAttemptInformation(logic, attemptIndex);

            if (m_outstream != nullptr) {
                std::wstringstream warningMessageStream;
                warningMessageStream
                    << L"Failed " << attemptInformation
                    << L"; ErrorCode:" << base::ErrorMessages::GetErrorCodeMessage(errorCode)
                    << L", ErrorMessage: " << e.GetErrorMessage();

                *m_outstream << warningMessageStream.str();
            }

            if (attemptIndex < m_policy->GetNumberOfAttempts() && IsAbsorbedException(logic, e)) {
                m_idleLogic->Run();

                continue;
            }

            std::wstringstream errorMessageStream;
            errorMessageStream
                << L"Retry Handler failed running: "
                << logicInformation;

            std::wstring errorMessage = errorMessageStream.str();

            if (m_outstream != nullptr) {
                *m_outstream << errorMessage;
            }

            throw RetryException(errorCode, errorMessage);
        }
    }
}

/**
 * Gets the information of a retry logic.
 */
std::wstring RetryHandler::GetLogicInformation(const IRetryLogic& logic) const
{
    std::wstringstream stream;
    stream << L"Logic: " << logic.GetName() << ", with Retry Policy: " << m_policy;
    return stream.str();
}

/**
 * Gets the retry logic attempt information.
 */
std::wstring RetryHandler::GetRetryLogicAttemptInformation(
    const IRetryLogic& logic,
    unsigned int attemptIndex) const
{
    std::wstringstream stream;

    stream
        << L"Attempt: " << attemptIndex << "/" << m_policy->GetNumberOfAttempts()
        << L" of Logic: " << logic;

    return stream.str();
}

/**
 * Checks whether an exception is absorbed.
 */
bool RetryHandler::IsAbsorbedException(
    const IRetryLogic& logic,
    base::BaseException& exception) const
{
    const std::vector<base::IClassTypeSharedPtr>& absorbedExceptions = logic.GetAbsorbedExceptions();

    for (const base::IClassTypeSharedPtr& classType : absorbedExceptions) {

        if (base::ClassType::IsAssignableFrom(exception)) {
            return true;
        }
    }

    return false;
}